File: Go/src/ray_old/plane.go
package main

type Plane struct {
	Normal  Vector
	Offset  float32
	surface Surface
}

func (p Plane) normal(v Vector) Vector {
	return p.Normal
}

func (p Plane) intersect(r Ray) (i Intersection, hit bool) {
	denom := p.Normal.Dot(r.direction)
	if denom > 0 {
		return Intersection{}, false
	} else {
		dist := (p.Normal.Dot(r.start) + p.Offset) / (-denom)
		return Intersection{thing: p, ray: r, distance: dist}, true
	}
}

func (p Plane) Surface() Surface {
	return p.surface
}

File: Kotlin/src/Plane.kt

data class Plane(val normal: Vector, val offset: Double, val surface: Surface)
    : Thing
{
    override fun normal(position: Vector) = this.normal
    override fun intersect(ray: Ray): Intersection? {
        val denom = this.normal.dot(ray.direction)
        return if (denom > 0) {
            null
        } else {
            val dist : Double = (this.normal.dot(ray.start) + this.offset) / (-denom)
            Intersection(this, ray, dist)
        }
    }

    override fun surface() = this.surface

}